File: vigenere/vigenere.py
from cs50 import get_string
import sys


def main():
    # Check input
    if not len(sys.argv) == 2:
        print("usage : python vigenere.py k")
        exit(1)
    if not sys.argv[1].isalpha():
        print("usage : python vigenere.py k")
        exit(1)

    # Get inputs
    plaintext = get_string("plaintext: ")
    key = sys.argv[1].upper()

    # Calculate cyphertext
    ciphertext = []
    travkey = 0
    for i in range(len(plaintext)):
        if plaintext[i].isalpha():
            ciphertext.append(determine(plaintext[i], ord(key[travkey % len(key)]) - 65))
            travkey += 1
        else:
            ciphertext.append(plaintext[i])
    print("ciphertext: " + ''.join(ciphertext))

    exit(0)


# Function to process char
def determine(letter, key):
    if letter.isupper():
        if ord(letter) + (key % 26) <= 90:
            return chr(ord(letter) + (key % 26))
        else:
            return chr(ord(letter) + (key % 26) - 26)
    elif ord(letter) + (key % 26) <= 122:
        return chr(ord(letter) + (key % 26))
    else:
        return chr(ord(letter) + (key % 26) - 26)


if __name__ == "__main__":
    main()

File: vigenere/vigenere.c
//Encrypts plaintext to cyphertext via a given key
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

// Receive command line arguments
int main(int argc, string argv[])
{

    string k;
    int q;

    // Check input is valid number of strings
    if (argc == 2)
    {
        //Check input is entirely alphabetical
        k = (argv[1]);
        q = 0;

        for (int h = 0; h < strlen(k); h++)
        {
            if (isalpha(k[h]))
            {
                q++;
            }
        }
    }

    // Check input meets criterea
    if (argc != 2 || q < strlen(k))
    {
        printf("usage : ./vigenere k \n");
        return 1;
    }

    //Prompt user for plaintext
    string p = get_string("plaintext: ");

    int o;

    // Start encryption process
    printf("ciphertext: ");

    // Operate on each letter individually
    for (int i = 0, n = 0; i < strlen(p); i++)
    {
        if (isalpha(p[i]))
        {
            // Check for key case
            if (isupper(k[n]))
            {
                o = (int)k[n] - 65;
            }
            else
            {
                o = (int)k[n] - 97;
            }
            // Check for plaintext case
            if (isupper(p[i]))
            {
                // Check to see whether 'wrap-around' is neseccary
                if ((p[i] + o) <= 'Z')
                {
                    printf("%c", p[i] + o);
                }
                else
                    // Count forwards from 'A'
                {
                    printf("%c", ((p[i] + o) - 90) + 64);
                }
            }
            else if (islower(p[i]))
            {
                // Same process for lower case
                if ((p[i] + o) <= 'z')
                {
                    printf("%c", p[i] + o);
                }
                else
                {
                    printf("%c", ((p[i] + o) - 122) + 96);
                }
            }
            n++;
        }
        // If char is not alphabetical
        else
        {
            printf("%c", p[i]);
        }
        // If exhausted length of key, start over again
        if (n >= strlen(k))
        {
            n = (n - strlen(k));
        }
    }
    // New line
    printf("\n");
    return 0;
}